cpp/include/cudf/hashing/detail/murmur_hash64.cuh

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cudf/detail/utilities/algorithm.cuh>
+#include <cudf/detail/utilities/hash_functions.cuh>
+
+#include <thrust/pair.h>
+
+namespace cudf {
+namespace hashing {
+namespace detail {
+
+// MurmurHash3_x64_128 implementation from
+// https://github.com/aappleby/smhasher/blob/master/src/MurmurHash3.cpp
+//-----------------------------------------------------------------------------
+// MurmurHash3_64_128 was written by Austin Appleby, and is placed in the public
+// domain. The author hereby disclaims copyright to this source code.
+// Note - The x86 and x64 versions do _not_ produce the same results, as the
+// algorithms are optimized for their respective platforms. You can still
+// compile and run any of them on any platform, but your performance with the
+// non-native version will be less than optimal.
+template <typename Key>
+struct MurmurHash3_64_128 {
+  using result_type = thrust::pair<uint64_t, uint64_t>;
+
+  constexpr MurmurHash3_64_128() = default;
+  constexpr MurmurHash3_64_128(uint64_t seed) : m_seed(seed) {}
+
+  __device__ inline uint32_t getblock32(std::byte const* data, cudf::size_type offset) const
+  {
+    // Read a 4-byte value from the data pointer as individual bytes for safe
+    // unaligned access (very likely for string types).
+    auto block = reinterpret_cast<uint8_t const*>(data + offset);
+    return block[0] | (block[1] << 8) | (block[2] << 16) | (block[3] << 24);
```

Comment:
does shift operator implicitly convert `uint8_t` to `uint32_t`?

Reply:
TIL: Integral promotions are performed on both operands.
The return type is the type of the left operand after integral promotions. 
Source: https://en.cppreference.com/w/cpp/language/operator_arithmetic#Bitwise_shift_operators


[prvalues](https://en.cppreference.com/w/cpp/language/value_category#prvalue) of small integral types (such as char) may be converted to prvalues of larger integral types (such as int). In particular, [arithmetic operators](https://en.cppreference.com/w/cpp/language/operator_arithmetic) do not accept types smaller than int as arguments, and integral promotions are automatically applied after lvalue-to-rvalue conversion, if applicable. This conversion always preserves the value.

Source: https://en.cppreference.com/w/cpp/language/implicit_conversion#Integral_promotion

So, converted to unsigned int.